sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -1500,6 +1500,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+ val V2_BUCKETING_SHUFFLE_ENABLED =
+    buildConf("spark.sql.sources.v2.bucketing.shuffle.enabled")
+      .doc("During a storage-partitioned join, whether to allow to shuffle only one side." +
+        "When only one side is KeyGroupedPartitioning, if the conditions are met, spark will " +
+        "only shuffle the other side. This optimization will reduce the amount of data that " +
```

Comment:
shall we make the algorithm smarter? If the other side is large, doing a `KeyGroupedPartitioning` may lead to skew and it's still better to shuffle both sides with hash partitioning.

Let's think of an extreme case: one side reports `KeyGroupedPartitioning` with only one partition, with this optimization, we end up with doing the join using a single thread.

Reply:
I think the `ShuffleSpec` "framework" in `EnsureRequirements` already takes this into consideration. This PR mainly makes `KeyGroupedShuffleSpec` behaves similar to `HashShuffleSpec` and be able to shuffle the other side (via making `canCreatePartitioning` return true).